### PR TITLE
XLAShardedTensor.to_local() support

### PR DESCRIPTION
**Pull request by Hoomaaan**
The implementation adds a to_local() method to XLAShardedTensor class that converts a sharded tensor back to its local representation while preserving gradient information.

1. Core Functionality:
- Returns the global tensor representation containing combined data across all devices
- Maintains the same device placement as the original XLAShardedTensor
- Creates a clone of the global tensor to ensure data independence
2. Gradient Handling:
- Preserves the requires_grad setting from the original tensor
- Maintains gradient values when converting to local representation
- Ensures proper gradient flow through the converted tensor
3. Test Coverage:
- Basic functionality test verifying shape and value preservation
- Dedicated gradient flow test ensuring:
requires_grad property is preserved
gradients are properly calculated and maintained
backward pass works correctly through the local tensor
gradient values are accurately preserved

### CODE REVIEW

File: test/spmd/test_xla_dtensor_to_local.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+class DTensorXLAFromLocalConversionTest(test_xla_sharding_base.XlaShardingTest):
+  """
+    Test suite for the automatic conversion of regular tensors to XLAShardedTensor
+    in DTensor.from_local() when using XLA device mesh.
+    """
+
+  @classmethod
+  def setUpClass(cls):
+    super().setUpClass()
```
Comment (bfolie) [inline test/spmd/test_xla_dtensor_to_local.py]: I don't think this method is necessary if there's no additional setup logic

File: test/spmd/test_xla_dtensor_to_local.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    # Call to local function
+    local_tensor = sharded_tensor.to_local()
+
+    # Verify requires_grad is preserved
+    self.assertTrue(local_tensor.requires_grad)
+
+    # All gradients should be 1.0 since we did a sum()
+    self.assertTrue(torch.allclose(local_tensor.grad, torch.ones_like(tensor)))
+
+    print("Gradient flow test successful")
```
Comment (bfolie) [inline test/spmd/test_xla_dtensor_to_local.py]: remove

File: test/spmd/test_xla_dtensor_to_local.py (hunk 1 above)
Comment (Hoomaaan) [reply test/spmd/test_xla_dtensor_to_local.py]: I added this to maintain consistency with other tests in the spmd folder, but happy to remove it if you think simplicity is more important than consistency here.

File: test/spmd/test_xla_dtensor_to_local.py (hunk 2 above)
Comment (Hoomaaan) [reply test/spmd/test_xla_dtensor_to_local.py]: Done!

File: test/spmd/test_xla_dtensor_to_local.py (hunk 1 above)
Comment (bfolie) [reply test/spmd/test_xla_dtensor_to_local.py]: There's no reason to redefine a method in a child class to simply call the parent class. It's just clutter, and consistency is not a good reason to propagate bad style.

Comment (Hoomaaan) [reply test/spmd/test_xla_dtensor_to_local.py]: @bfolie I agree. Removed the constructor.

**Review by bfolie — APPROVED** [review]

File: torch_xla/distributed/spmd/xla_sharded_tensor.py [hunk 3]
```diff
@@ -150,6 +150,29 @@ def load_local_shards_(self, shards: List[XLAShard]):
...
+    Returns:
+        torch.Tensor: The global tensor representation with appropriate requires_grad setting.
+    """
+
+    # Create a new tensor with the same values of global_tensor
+    result = self.global_tensor.clone()
+    # Since global tensor is detached, add requires_grad and grad values back to the local tensor
+    if self.requires_grad:
+      result.requires_grad = self.requires_grad
+      result.grad = self.grad
```
Comment (rpsilva-aws) [inline torch_xla/distributed/spmd/xla_sharded_tensor.py]: The grad doesn't need to be cloned? Is it fine that we break the reference for the tensor, but not its grad? See if we can add a test that shows it is the case (e.g. both the prior and newer tensor updating the same grad).

File: torch_xla/distributed/spmd/xla_sharded_tensor.py [hunk 4]
```diff
@@ -150,6 +150,29 @@ def load_local_shards_(self, shards: List[XLAShard]):
...
+    
+    Returns:
+        torch.Tensor: The global tensor representation with appropriate requires_grad setting.
+    """
+
+    # Create a new tensor with the same values of global_tensor
+    result = self.global_tensor.clone()
+    # Since global tensor is detached, add requires_grad and grad values back to the local tensor
+    if self.requires_grad:
+      result.requires_grad = self.requires_grad
```
Comment (rpsilva-aws) [inline torch_xla/distributed/spmd/xla_sharded_tensor.py]: nit: result.requires_grad_(self.requires_grad) for in-place.

**Review by rpsilva-aws — CHANGES_REQUESTED** [review]

File: torch_xla/distributed/spmd/xla_sharded_tensor.py (hunk 3 above)
Comment (Hoomaaan) [reply torch_xla/distributed/spmd/xla_sharded_tensor.py]: Updated to clone the self.grad if available

File: torch_xla/distributed/spmd/xla_sharded_tensor.py (hunk 4 above)
Comment (Hoomaaan) [reply torch_xla/distributed/spmd/xla_sharded_tensor.py]: Done!
